### PR TITLE
Adds asset_group_name property to InputContext, OutputContext

### PR DESCRIPTION
**Pull request by cmpadden**
## Summary & Motivation

- Closes https://github.com/dagster-io/dagster/issues/13945
- Adds ability to get input / output asset group name from context in I/O manager

## How I Tested These Changes

- `pytest`

### CODE REVIEW

File: python_modules/dagster/dagster_tests/storage_tests/test_io_manager.py
Comment (jamiedemaria) [inline python_modules/dagster/dagster_tests/storage_tests/test_io_manager.py:759]: can you add a test for assets with no group name set?

**Review by jamiedemaria — APPROVED** [review]
one test that would be good to add, but lgtm

**Comment by sryza** [comment]
Pending this discussion, I wonder if we should expose `spec` instead of the asset properties individually? https://github.com/dagster-io/internal/discussions/9837

File: python_modules/dagster/dagster_tests/storage_tests/test_io_manager.py
Comment (cmpadden) [reply python_modules/dagster/dagster_tests/storage_tests/test_io_manager.py]: I've added the test asserting group is _default_, thanks.

**Comment by cmpadden** [comment]
> Pending this discussion, I wonder if we should expose `spec` instead of the asset properties individually? [dagster-io/internal#9837](https://github.com/dagster-io/internal/discussions/9837)

Jamie and I discussed that a bit in Slack, and I think we're on the same page that exposing the spec would be a more flexible and maintainable path moving forward. I'd be happy to expose that here once it's available. In the meantime do you think we should merge `asset_group_names` or hold off until the specs PRs are merged? Either way works for me.

**Comment by jamiedemaria** [comment]
I'm very much in favor of the `AssetSpec` approach. my two questions:
@cmpadden what's the urgency of this PR? is it just a nice to have or are we seeing a lot of user requests?
@sryza what's your thinking on time frame of exposing `AssetSpec`? on the scale of weeks or months?

I think if this is a relatively low priority feature to add and we think we can expose `AssetSpec` in the near term, I'd say let's wait for `AssetSpec`

**Comment by jamiedemaria** [comment]
ok! In that case I definitely say let's hold off

**Review by jamiedemaria — CHANGES_REQUESTED** [review]
switching to requesting changes to prevent accidental merging

**Pull request closed without merging**
